packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp

```diff
@@ -0,0 +1,499 @@
+              regQueue.push(art_pt);
+              regQueue.push(outs[i]);
+              break;
+            }
+          }
+          
+        }
+        femv->endFill();
+      }
+      
+      //call this->propagate, which, per-processor, finds one bcc at a time.
+      int* t = propagate();
+      bcc_count++;
+      delete [] t;
+      //check for OWNED articulation points
+      for(int i = 0; i < nLocalOwned; i++){
+        if(femvData[i].getGroundingStatus() == ICEPROPGS_FULL){
+          lno_t out_degree = g->out_degree(i);
+          lno_t* outs = g->out_vertices(i);
+          for(int j = 0; j < out_degree; j++){
+            if(femvData[outs[j]].getGroundingStatus() < ICEPROPGS_FULL){
+              art_queue.push(i);
+              articulation_point_flags[i] = 1;
+              break;
+            }
+          }
+        }
+      }
+      //clear half labels
+      for(int i = 0; i < nLocalOwned+nLocalCopy; i++){
+        label_t label = femvData[i];
+        if(label.getGroundingStatus() == ICEPROPGS_HALF){
+          label.first_label = -1;
+          label.first_sender = -1;
+          label.bcc_name = -1;
+          femv->switchActiveMultiVector();
+          femv->replaceLocalValue(i,0,label);
+          femv->switchActiveMultiVector();
+        }
+      }
+      //pop articulation points off of the art_queue if necessary.
+      if(!art_queue.empty()){
+        bool pop_art = true;
+        while(pop_art && !art_queue.empty()){
+          lno_t top_art_degree = g->out_degree(art_queue.front());
+          lno_t* art_outs = g->out_vertices(g,art_queue.front());
+        
+          for(int i = 0; i < top_art_degree; i++){
+            if(femvData[art_outs[i]].getGroundingStatus() < ICEPROPGS_FULL){
+              pop_art = false;
+            }
+          }
+          if(pop_art) art_queue.pop();
+        }
+      }
+    }
+    for(int i = 0; i < nLocalOwned; i++){
+      if(!articulation_point_flags[i]){
+        label_t label = femvData[i];
+        label.is_art = false;
+        femv->replaceLocalValue(i,0,label);
+      }
+    }
+    return femv->getData(0);
+  }
+
+private:
+  int me; 	    //my processor rank
+  int np;	    //number of processors
+  int nLocalOwned;  //number of vertices owned by this processor
+  int nLocalCopy;   //number of copies of off-processor vertices on this proc
+  int nVec;	    //number of vectors in multivector
+  
+  const Teuchos::RCP<const Teuchos::Comm<int> > comm; //MPI communicator
+  icePropGraph<lno_t>* g;	    //csr representation of vertices on this processor
+
+  Teuchos::RCP<const MAP> mapOwned;       //Tpetra::Map including only owned
+  Teuchos::RCP<const MAP> mapWithCopies;  //Tpetra::Map including owned
+                                            //vertices and copies
+
+  Teuchos::RCP<femv_t> femv;
+  
+  Teuchos::ArrayRCP<const scalar_t> femvData;
+};
+
+//std::queue<iceSheetPropagation<Tpetra::Map<>>::lno_t> iceSheetPropagation<Tpetra::Map<>> ::regQueue;
```

Comment:
Delete these comments; the queues were moved.

Reply:
Done.